### PR TITLE
feat: allow setting a service type and annotations for a pooler resource

### LINKED ISSUE #3141: [Feature]: Allow exposing a pooler resource externally using a loadbalancer
### Is there an existing issue already for this feature request/idea?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new feature request to be evaluated.

### What problem is this feature going to solve? Why should it be added?

In our use-case, CNPG runs in a specific K8S cluster while apps might be on a different K8S cluster. Hence, it is necessary to expose the pooler service externally using a service type `LoadBalancer`. Other people might have a similar use-case and will benefit from having this built into CNPG.

### Describe the solution you'd like

I am proposing that the `Pooler` resource be modified to specify a flag that denotes if the pooler should be exposed. Also, add a new field `loadBalancerSourceRanges` that is then used to create the pooler service.

```
apiVersion: postgresql.cnpg.io/v1
kind: Pooler
metadata:
  name: pooler-example-rw
spec:
  cluster:
    name: cluster-example
  instances: 1
  type: rw
  external: true
  loadBalancerSourceRanges:
  - 192.168.122.0/24
  - 0.0.0.0/32
  pgbouncer:
    poolMode: transaction
```

### Describe alternatives you've considered

So far we created a new service of type loadbalancer to expose the pooler externally. While this works fine, this introduces complexity in managing the deployment with two services, one managed by CNPG and one not managed by CNPG. It will be nice to have everything in the same umbrella.

### Additional context

_No response_

### Backport?

No

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by achanda**
Closes #3141

### CODE REVIEW

**Comment by achanda** [comment]
Todo:

- [ ] Make annotations optional
- [ ] More tests
- [ ] Docs update

File: api/v1/pooler_types.go
Comment (phisco) [inline api/v1/pooler_types.go:86]: Ideally this could fall under a serviceTemplate, providing a full service spec and metadata for labels and annotations, we could still apply the labels/annotation inheritance, but allow setting service specific ones here.

Comment (armru) [reply api/v1/pooler_types.go:86]: I agree

File: pkg/servicespec/builder.go
Comment (phisco) [inline pkg/servicespec/builder.go:26]:
```suggestion
// Builder enables users to create a serviceTemplate starting from a baseline
```

File: pkg/servicespec/builder.go
Comment (phisco) [inline pkg/servicespec/builder.go:32]:
```suggestion
// New creates a new empty serviceTemplate builder
```

File: pkg/servicespec/builder.go
Comment (phisco) [inline pkg/servicespec/builder.go:75]:
```suggestion
// Build gets the final ServiceTemplate
```

File: pkg/specs/pgbouncer/services.go
Comment (phisco) [inline pkg/specs/pgbouncer/services.go:46]: Why not moving all these to the builder too and only setting that here?

**Review by phisco — COMMENTED** [review]
Awesome work @achanda! I would get a little bit further with the templating before merging!🙏

File: pkg/specs/pgbouncer/services.go
Comment (achanda) [reply pkg/specs/pgbouncer/services.go]: Thanks, done that now. I also included the copy paste fixes you suggested. Once we agree on the implementation, I will work on tests.

File: pkg/servicespec/builder.go
Comment (phisco) [inline pkg/servicespec/builder.go:79]: Here it should get a whole ServicePort, iterate over the provided ports, if any, and if a port with the given name is found, update it, otherwise append it at the end of the list.

File: pkg/servicespec/builder.go
Comment (phisco) [inline pkg/servicespec/builder.go:92]: Here it should get a bool and overwrite only if true, only adding it otherwise.

**Comment by achanda** [comment]
@phisco what usecase do you have in mind to enable adding to selectors and ports? I was thinking we don't enable adding to these and keep these internal. Later, if there is a usecase for modifying this then we revisit.

**Comment by phisco** [comment]
@achanda just for the sake of completeness, as long as we handle that properly, I don't see a reason not to allow that right away. Not a strong requirement though.

**Comment by achanda** [comment]
@phisco done and pushed, PTAL 🙏

**Comment by phisco** [comment]
Awesome! Thanks, @achanda! The code looks good, I'll give it a spin tomorrow 🙏

File: pkg/specs/pgbouncer/services.go
Comment (phisco) [inline pkg/specs/pgbouncer/services.go:37]: Here you should be setting the "default" service type instead, as the one from `pooler.Spec.ServiceTemplate.Spec.Type` is already set by `NewFrom` implicitly.
Given that you don't want to overwrite the default value, you should also add a boolean `overwrite` parameter, preferably as last parameter to adhere to the standard we already have for the podspec builder.
```suggestion
		WithServiceType(corev1.ClusterIP, true).
```

File: pkg/specs/pgbouncer/services.go
Comment (phisco) [inline pkg/specs/pgbouncer/services.go:44]: Please invert the two parameters to have the `overwrite` one last.
```suggestion
		WithSelector(pooler.Name, true).
```

File: config/crd/bases/postgresql.cnpg.io_poolers.yaml
Comment (phisco) [inline config/crd/bases/postgresql.cnpg.io_poolers.yaml:1]: unfortunately it looks like this CRD became too long and can not be applied anymore as the `kubectl.kubernetes.io/last-applied-configuration` annotation `kubectl apply` adds automatically, would not fit anymore in the maximum allowed size for annotations, at least on Kubernetes `v1.28.0`:
```
The CustomResourceDefinition "poolers.postgresql.cnpg.io" is invalid: metadata.annotations: Too long: must have at most 262144 bytes
```
I'll have to think about a workaround.

**Review by phisco — CHANGES_REQUESTED** [review]
can you run a `make reviewable` to be sure the linter is happy and that the unit tests are all green too?

File: config/crd/bases/postgresql.cnpg.io_poolers.yaml
Comment (phisco) [reply config/crd/bases/postgresql.cnpg.io_poolers.yaml:1]: Unfortunately that's a pretty difficult issue to solve properly for an embedded resource :( https://book.kubebuilder.io/faq#the-error-too-long-must-have-at-most-262144-bytes-is-faced-when-i-run-make-install-to-apply-the-crd-manifests-how-to-solve-it-why-this-error-is-faced.

The best option would be to drop all descriptions for the fields in the service (and pod) template, but unfortunately I didn't find any available tool to do that. Usually you don't care about required fields either, so such tool would have to be able to "vendor" a struct from an upstream project, stripping annotations. We have a similar need in Crossplane, so I could take the chance to develop something able to do that. But that will require time. In the meantime we could either hack something together to cleanup the crds after having generated them using something like `yq` for example 🤔

Comment (achanda) [reply config/crd/bases/postgresql.cnpg.io_poolers.yaml:1]: I need to figure this one out, I will get to this next week.

**Comment by achanda** [comment]
@phisco I checked that I can apply these CRDs on a kind cluster. This should be ready for review now.

```
╭─Abhishek@AbhishekMacbookPro ~/src/cloudnative-pg ‹external-pooler›
╰─$ kubectl apply -f config/crd/bases/
customresourcedefinition.apiextensions.k8s.io/backups.postgresql.cnpg.io created
customresourcedefinition.apiextensions.k8s.io/clusters.postgresql.cnpg.io unchanged
customresourcedefinition.apiextensions.k8s.io/poolers.postgresql.cnpg.io created
customresourcedefinition.apiextensions.k8s.io/scheduledbackups.postgresql.cnpg.io created
```

**Comment by phisco** [comment]
Depending on the decision taken by https://github.com/cloudnative-pg/cloudnative-pg/issues/3697 we could be fine with the full CRD for the pooler too here.

**Comment by phisco** [comment]
@achanda, as you can see we agreed to use server-side apply for CRDs: https://github.com/cloudnative-pg/cloudnative-pg/pull/3729

This means here we can have the full schema produced, could you revert the last commit? Sorry for the back and forth, but it was a big decision to take.

**Comment by achanda** [comment]
No worries @phisco I have updated this branch now. Thanks for all the help 🙏

File: api/v1/pooler_types.go
Comment (phisco) [inline api/v1/pooler_types.go:139]:
```suggestion
	ObjectMeta Metadata `json:"metadata,omitempty"`
```
Same as the pod template above, will need to be converted

File: pkg/specs/pgbouncer/services.go
Comment (phisco) [inline pkg/specs/pgbouncer/services.go:36]:
```suggestion
        poolerHash, err := hash.ComputeVersionedHash(pooler.Spec, 3)
	if err != nil {
		return nil, err
	}

	serviceTemplate := servicespec.NewFrom(pooler.Spec.ServiceTemplate).
		WithLabel(utils.PgbouncerNameLabel, pooler.Name).
		WithLabel(utils.ClusterLabelName, cluster.Name).
		WithLabel(utils.PodRoleLabelName, string(utils.PodRolePooler)).
		WithAnnotation(utils.PoolerSpecHashAnnotationName, poolerHash).
```

File: pkg/specs/pgbouncer/services.go
Comment (phisco) [inline pkg/specs/pgbouncer/services.go:63]:
```suggestion
			Annotations: serviceTemplate.ObjectMeta.Annotations,
```

Comment (phisco) [reply pkg/specs/pgbouncer/services.go:63]: Same should be done for the Labels above, github didn't let me:
```
			Labels:      serviceTemplate.ObjectMeta.Labels,
```

File: pkg/specs/pgbouncer/services.go
Comment (phisco) [inline pkg/specs/pgbouncer/services.go:51]: moved above
```suggestion
```

**Review by phisco — CHANGES_REQUESTED** [review]

**Comment by achanda** [comment]
@phisco gentle ping on this, do you need anything else from me?

File: pkg/servicespec/builder.go
Comment (phisco) [inline pkg/servicespec/builder.go:72]: This should set the type also if it's not set:
```suggestion
	if overwrite || builder.status.Spec.Type == "" {
```

File: pkg/servicespec/builder.go
Comment (phisco) [inline pkg/servicespec/builder.go:99]: Same as above, overwrite means if present overwrite:
```suggestion
func (builder *Builder) WithSelector(name string, overwrite bool) *Builder {
	s := builder.status.Spec.Selector

	if len(s) == 0 {
		s = make(map[string]string, 1)
	}

	if _, exists := s[utils.PgbouncerNameLabel]; overwrite || !exists {
		s[utils.PgbouncerNameLabel] = name
	}

	builder.status.Spec.Selector = s

	return builder
}
```

**Review by phisco — APPROVED** [review]
Thanks @achanda! Sorry for taking so long, I had a pretty busy period at work 😅

**Comment by achanda** [comment]
@phisco no worries at all, I know last week was kubecon :) The failing tests look unrelated to me, mind running the tests again please? I don't have permission

**Comment by phisco** [comment]
Yes, I think they are too, I'll let the second reviewer decide if they want to run them again or just mark it as good to go straight away 👍

**Comment by sxd** [comment]
Hi @achanda would you like to add more documentation? I just added a small example but maybe you want to create a file in docs/src/samples/ directory that can be used ?

**Comment by achanda** [comment]
@sxd done now, added a new file named pooler-external.yaml and referenced that in the doc

File: docs/src/connection_pooling.md
Comment (sxd) [inline docs/src/connection_pooling.md:287]: there's a full section of samples in samples.md where this should go

Comment (achanda) [reply docs/src/connection_pooling.md]: Made that change, thanks!

**Review by sxd — APPROVED** [review]

**Pull request merged**
